executor/aggfuncs/func_max_min.go

```diff
@@ -14,49 +14,174 @@
 package aggfuncs
 
 import (
+	"sort"
+
 	"github.com/pingcap/tidb/sessionctx"
 	"github.com/pingcap/tidb/types"
 	"github.com/pingcap/tidb/types/json"
 	"github.com/pingcap/tidb/util/chunk"
 	"github.com/pingcap/tidb/util/stringutil"
 )
 
+type maxMinQueue struct {
+	queue   []interface{}
+	varSet  map[interface{}]int64
+	isMax   bool
+	dirty   bool
+	cmpFunc func(i, j interface{}) int
+}
+
+func newMaxMinQueue(isMax bool) *maxMinQueue {
+	return &maxMinQueue{
+		queue:  make([]interface{}, 0),
+		varSet: make(map[interface{}]int64),
+		isMax:  isMax,
+	}
+}
+
+func (m *maxMinQueue) Resort() {
+	if len(m.queue) == 0 {
+		return
+	}
+	sort.Slice(m.queue, func(i, j int) bool {
+		if m.isMax {
+			return m.cmpFunc(m.queue[i], m.queue[j]) > 0
+		}
+		return m.cmpFunc(m.queue[i], m.queue[j]) < 0
+	})
+}
+
+func (m *maxMinQueue) Top() (val interface{}, isEmpty bool) {
+	if len(m.queue) == 0 {
+		return nil, true
+	}
+	if m.dirty {
+		m.Resort()
```

Comment:
We need an algorithm that `Top` takes O(1) time complex whether it's dirty or not, so the `Resort` can not be introduced in this situation.

It's better to use a heap plus a map to solve this problem.
I wrote a demo code for you.

```go
type maxMinQueue struct {
	h heap.Interface
	count map[string]int
}

func (m *maxMinQueue) push(val string) {
	m.count[val]++
	if m.count[val] == 1 {
		heap.Push(m.h, val)
	}
}

func (m *maxMinQueue) del(val string) {
	if m.count[val] == 1 {
		delete(m.count, val)
	}
	m.count[val]--
}

func (m *maxMinQueue) top() string {
retry:
	t := heap.Pop(m.h).(string)
	if m.count[t] == 0 {
		// Even there's a retry here, we can assume that There're at most N times call of top
		//  if there're N times call of push. So the time complex of top can be O(1).
		goto retry
	}
	return t
}
```

Reply:
```
name                                                                                                                                sort time/op  heap time/op  delta
WindowFunctionsWithSlidingWindow/(func:max,_aggColType:int(11),_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)-16     18.7ms ± 2%  21.0ms ± 0%  +12.73%  (p=0.000 n=10+8)
WindowFunctionsWithSlidingWindow/(func:max,_aggColType:float,_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)-16       18.6ms ± 1%  21.1ms ± 1%  +13.28%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:min,_aggColType:int(11),_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)-16     18.7ms ± 1%  20.6ms ± 2%  +10.49%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:min,_aggColType:float,_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)-16       18.6ms ± 1%  20.6ms ± 1%  +11.10%  (p=0.000 n=8+10)
WindowFunctionsWithSlidingWindow/(func:max,_aggColType:int(11),_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)#01-16  18.4ms ± 1%  18.6ms ± 1%   +1.04%  (p=0.000 n=10+10)
WindowFunctionsWithSlidingWindow/(func:max,_aggColType:float,_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)#01-16    18.3ms ± 0%  18.7ms ± 1%   +2.12%  (p=0.000 n=9+10)
WindowFunctionsWithSlidingWindow/(func:min,_aggColType:int(11),_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)#01-16  18.5ms ± 1%  18.6ms ± 3%     ~     (p=0.360 n=8+10)
WindowFunctionsWithSlidingWindow/(func:min,_aggColType:float,_numFunc:1,_ndv:100,_rows:100000,_sorted:true,_concurrency:1)#01-16    18.4ms ± 0%  19.1ms ± 3%   +3.64%  (p=0.000 n=8+10)
```

Reply:
Using heap seem slower than sort. Maybe I was using it the wrong way?

Reply:
I used it wrong before.